### PR TITLE
Feature/presentation mode

### PR DESCRIPTION
**Pull request by cuginoAle**

### CODE REVIEW

**Comment by mcollina** [comment]
Can you please include a screenshot?

**Comment by cuginoAle** [comment]
Screenshots:

normal mode
screenshot 2018-11-14 at 12 08 32


presentation mode
screenshot 2018-11-14 at 12 08 43

**Comment by mcollina** [comment]
I don't think this will do.

There are few issues with flame design on projectors:

1. the frames are not clearly distinguishable, as the white/grey line between then it's so thin that some projectors have trouble in showing it
2. the frames text is not contrasted enough vs the background, and it's almost impossible to read from a distance.

Maybe we should make these changes even on normal screens, they won't hurt (no presentation mode).

**Comment by cuginoAle** [comment]
I am still working on it ;) [wip]
I'll let you know when it's ready!

**Comment by cuginoAle** [comment]
- Thicker lines
- Taller frames
- Higher contrast 

Please find an [updated demo here](https://upload.clinicjs.org/public/2d36206e063e1fe5ef9f64d2277a360c10d084986f480718acfb9efb42c10073/5759.clinic-flame.html) (the presentation mode switch is in the Options drop-down) :

**Comment by cuginoAle** [comment]
This PR introduces some changes on how the frame [size/position gets calculated](https://github.com/nearform/node-clinic-flame/blob/c4f5ecbaa482300098511b484aef53c910db9f21/visualizer/flame-graph-frame.js#L35)
I double-checked no side effects have been introduced by this changes but please confirm everything is ok.
This also fixes a tiny bug (that becomes more evident when in PresentationMode) where the frame bottom line was overlapped by the frame immediately below it on hover.
Please notice how the bottom line colour switches between blue and grey and how the right border of  `app.get` gets whiter over time.
ezgif com-video-to-gif

**Comment by goto-bus-stop** [comment]
IIRC in the call we discussed that an environment variable might be ideal, so presenters can do `export CLINIC_FLAME_PRESENTER=1` or something once and then just use the commands as normal, so they don't have to explain that workshoppers/listeners should omit `--presentation-mode`. having a `presentationMode` option is still fine I think but we could default it to `!!process.env.CLINIC_FLAME_PRESENTER` or so maybe?

**Comment by cuginoAle** [comment]
After the latest update to enable `presentation mode` by default we need to set the `PRESENTATION_MODE` env variable to true:
`export PRESENTATION_MODE=true`

**Comment by goto-bus-stop** [comment]
code wise it LGTM! i have no presenting experience and don't know how things look on beamers so i won't/can't give full approval on that front :laughing:

**Comment by mcollina** [comment]
Can you add an updated example?

One feedback is that in the given case, when clicking on the "hottest frame" the top of the stack is not visible. I think we should also center vertically on the given frame.

**Comment by mcollina** [comment]
@AlanSl can you please validate this on your testing beamer?

**Comment by cuginoAle** [comment]
> Can you add an updated example?
> 
> One feedback is that in the given case, when clicking on the "hottest frame" the top of the stack is not visible. I think we should also center vertically on the given frame.

This is already [implemented here](https://github.com/nearform/node-clinic-flame/pull/29) ;)
I'll merge that PR shortly and upload an updated example

**Comment by cuginoAle** [comment]
@mcollina Please find an example here: https://upload.clinicjs.org/public/d355f992114ab6a43a1adbc17bb56ad731a324e9c806d62d8b141ec15d8e45e4/5759.clinic-flame.html#selectedNode=933&zoomedNode=&exclude=3e&merged=true

**Comment by AlanSl** [comment]
I've done quite a bit of testing and tinkering on this using a projector. Challenge # 1 is creating test conditions worse than anything we might see in a real presentation room. I've deliberately gone for very extreme conditions here since every time I've tested before, some presentation room has given us a setup that is much worse... So this is about as bad as it can get, projected on a sheet opposite a window letting in daylight:

image

This PR in presentation mode is better, but still not very clear under extreme conditions. In particular, the UI elements of the toolbar become very unclear, and the flamegraph shape and text still doesn't have quite enough contrast:

(note - the blurriness in the bottom right corner in this and the next photo is because the camera can't focus on the whole thing, it's sharper in real life)

image

I've tinkered with the settings and pushed some commits to a branch building on this one called `pres-mode-adjust`, here's a photo with the adjustments. You can see the shape of the flamegraph even in these extreme conditions now, and the text is clearer and the colours a little more distinct:

(note that like the previous photo, the blurriness in the bottom right corner is because of the camera focus, it's sharper in real life)

image

**Comment by AlanSl** [comment]
Thinking about it, there's enough commits in my adjustments branch that it's more like a PR of its own than something that could be added to an existing PR, so probably the best way to proceed is if we merge this as-is but don't release it yet, then I'll post the amendments as a another PR and we release them both in the same minor release.

**Comment by mcollina** [comment]
+1

File: visualizer/options-menu.js [hunk 1]
```diff
@@ -61,6 +79,37 @@ class OptionsMenu extends HtmlContent {
...
+              <input id='${d.id}' type='checkbox' "${d.value ? 'checked' : ''}" /> 
+              <span class='icon-wrapper'>
+                <img class="icon-img checked" data-inline-svg src="/visualizer/assets/icons/checkbox-checked.svg" />
+                <img class="icon-img unchecked" data-inline-svg src="/visualizer/assets/icons/checkbox-unchecked.svg" />
+              </span>
+              <span class='copy-wrapper'>${d.title}</span>
+            </label>
+          </div>
+        `)
+      })
```
Comment (goto-bus-stop) [inline visualizer/options-menu.js]: I think this is basically identical to the tree generated by `addFgOptionCheckbox`, only appended to a different parent element. Maybe `addFgOptionCheckbox` could be reused here with some slight tweaks? eg. get pass in the parent as an option or something

**Comment by AlanSl** [comment]
LGTM

**Pull request merged**